android/shared/src/main/kotlin/kmp/android/shared/navigation/Destination.kt

```diff
@@ -0,0 +1,149 @@
+import com.google.accompanist.navigation.material.ExperimentalMaterialNavigationApi
+import com.google.accompanist.navigation.material.bottomSheet
+
+/**
+ * Base class for navigation graph root class.
+ */
+interface RootDestination {
+    val route: String
+}
+
+/**
+ * Base class for any destination the app uses.
+ */
+abstract class Destination(parent: RootDestination?) {
+    /**
+     * The root id of the destination graph used in the route <root>/<destinationId>/<argument>...
+     * In case the parent is null, we do not need any prefix and use the destinationId by itself <destinationId>/<argument>...
+     */
+    private val root: String = parent?.route?.let { "$it/" } ?: ""
+
+    /**
+     * The id of the destination, without root id and nav arguments.
+     */
+    protected abstract val destinationId: String
+
+    open val arguments: List<NamedNavArgument> = emptyList()
+
+    /**
+     * Route used to determine this destination from others. Do not use for navigating.
+     */
+    val route
+        get() = constructRoute()
+
+    /**
+     * Creates a navigate-able route with the [arguments] provided.
+     * @param arguments Arguments that should be used for this destination. The number and order of arguments
+     * must exactly match the number and order of arguments this destination requires. For a default value, use null.
+     */
+    operator fun invoke(vararg arguments: Any?) = constructRouteForNavigation(*arguments)
+
+    /**
+     * Constructs a route with argument names in the following format:
+     * - e.g. Destination called `detail` with arguments named `id` and `name`:
+     *      "detail?id={id}&name={name}"
+     * - e.g. Destination called `home` with no arguments:
+     *      "home"
+     */
+    private fun constructRoute() = createUri(
+        path = "$root$destinationId",
+        argList = arguments,
+        key = { it.name },
+        value = { "{${it.name}}" },
+    )
+
+    /**
+     * Constructs a route with argument names in the following format:
+     * - e.g. Destination called `detail` with arguments named `id` = 5 and `name` = "test":
+     *      "detail?id=5&name=test"
+     * - e.g. Destination called `home` with no arguments:
+     *      "home"
+     */
+    private fun constructRouteForNavigation(
+        vararg routeArguments: Any?,
+    ): String {
+        require(routeArguments.size == arguments.size) {
+            "The routeArgument count must match this destination argument count.\n" +
+                "If needed, pass null for default value of argument."
+        }
+        val args = routeArguments.zip(arguments).map { (routeArg, arg) ->
+            // Provided arg or default arg
+            val value = (routeArg ?: arg.argument.defaultValue)
+
+            require(value != null || arg.argument.isNullable) {
+                "The argument ${arg.name} is null as well as it's default value, but it was not marked as nullable."
+            }
```

Comment:
@vojta-horanek I added this to enable nullable arguments, in the previous version when I used null it converted it to string and returned string null instead of null value 😄